File: src/common/wasm/hypot.ts
/**
 * WASM-accelerated element-wise hypot.
 *
 * Binary: out[i] = hypot(a[i], b[i])  (same-shape contiguous arrays)
 * Scalar: out[i] = hypot(a[i], scalar)
 * Returns null if WASM can't handle this case.
 * Float types use native kernels; int64/uint64 use native i64 kernels
 * that convert to f64 in WASM (avoiding JS BigInt->Number overhead).
 */

import {
  hypot_f64,
  hypot_f32,
  hypot_scalar_f64,
  hypot_scalar_f32,
  hypot_i64,
  hypot_scalar_i64,
  hypot_i32,
  hypot_scalar_i32,
  hypot_i16,
  hypot_scalar_i16,
  hypot_i8,
  hypot_scalar_i8,
} from './bins/hypot.wasm';
import {
  wasmMalloc,
  resetScratchAllocator,
  resolveInputPtr,
  f16InputToScratchF32,
  f32OutputToF16Region,
} from './runtime';
import { ArrayStorage } from '../storage';
import { promoteDTypes, type DType, type TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 64;

type BinaryFn = (aPtr: number, bPtr: number, outPtr: number, N: number) => void;
type ScalarFn = (aPtr: number, outPtr: number, N: number, scalar: number) => void;

const binaryKernels: Partial<Record<DType, BinaryFn>> = {
  float64: hypot_f64,
  float32: hypot_f32,
  float16: hypot_f32,
};

const scalarKernels: Partial<Record<DType, ScalarFn>> = {
  float64: hypot_scalar_f64,
  float32: hypot_scalar_f32,
  float16: hypot_scalar_f32,
};

// Int64 kernels: native int input, f64 output
const intBinaryKernels: Partial<Record<DType, BinaryFn>> = {
  int64: hypot_i64,
  uint64: hypot_i64,
  int32: hypot_i32,
  uint32: hypot_i32,
  int16: hypot_i16,
  uint16: hypot_i16,
  int8: hypot_i8,
  uint8: hypot_i8,
};

const intScalarKernels: Partial<Record<DType, ScalarFn>> = {
  int64: hypot_scalar_i64,
  uint64: hypot_scalar_i64,
  int32: hypot_scalar_i32,
  uint32: hypot_scalar_i32,
  int16: hypot_scalar_i16,
  uint16: hypot_scalar_i16,
  int8: hypot_scalar_i8,
  uint8: hypot_scalar_i8,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
  float16: Float32Array,
  int64: BigInt64Array,
  uint64: BigUint64Array,
  int32: Int32Array,
  uint32: Uint32Array,
  int16: Int16Array,
  uint16: Uint16Array,
  int8: Int8Array,
  uint8: Uint8Array,
};

/**
 * WASM-accelerated element-wise hypot of two same-shape contiguous arrays.
 * Returns null if WASM can't handle.
 */
export function wasmHypot(a: ArrayStorage, b: ArrayStorage): ArrayStorage | null {
  if (!a.isCContiguous || !b.isCContiguous) return null;

  // WASM kernel does not broadcast — sizes must match
  if (a.size !== b.size) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = promoteDTypes(a.dtype, b.dtype);

  // Float path
  const floatKernel = binaryKernels[dtype];
  if (floatKernel) {
    const Ctor = ctorMap[dtype]!;
    const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
    const outBytes = size * bpe;

    const outRegion = wasmMalloc(outBytes);
    if (!outRegion) return null;

    wasmConfig.wasmCallCount++;

    resetScratchAllocator();

    let aPtr: number;
    let bPtr: number;
    if (dtype === 'float16') {
      aPtr = f16InputToScratchF32(a, size);
      bPtr = f16InputToScratchF32(b, size);
    } else {
      aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, bpe);
      bPtr = resolveInputPtr(b.data, b.isWasmBacked, b.wasmPtr, b.offset, size, bpe);
    }

    floatKernel(aPtr, bPtr, outRegion.ptr, size);

    if (dtype === 'float16') {
      const f16Region = f32OutputToF16Region(outRegion, size);
      outRegion.release();
      if (!f16Region) return null;
      return ArrayStorage.fromWasmRegion(
        Array.from(a.shape),
        dtype,
        f16Region,
        size,
        Float16Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
      );
    }

    return ArrayStorage.fromWasmRegion(
      Array.from(a.shape),
      dtype,
      outRegion,
      size,
      Ctor as unknown as new (buffer: ArrayBuffer, byteOffset: number, length: number) => TypedArray
    );
  }

  // Int path: native int input, f64 output
  const intKernel = intBinaryKernels[dtype];
  if (intKernel) {
    const InputCtor = ctorMap[dtype]!;
    const inputBpe = (InputCtor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
    const outBytes = size * 8;

    const outRegion = wasmMalloc(outBytes);
    if (!outRegion) return null;

    wasmConfig.wasmCallCount++;

    resetScratchAllocator();
    const aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, inputBpe);
    const bPtr = resolveInputPtr(b.data, b.isWasmBacked, b.wasmPtr, b.offset, size, inputBpe);

    intKernel(aPtr, bPtr, outRegion.ptr, size);

    return ArrayStorage.fromWasmRegion(
      Array.from(a.shape),
      'float64',
      outRegion,
      size,
      Float64Array as unknown as new (
        buffer: ArrayBuffer,
        byteOffset: number,
        length: number
      ) => TypedArray
    );
  }

  return null;
}

/**
 * WASM-accelerated element-wise hypot with scalar.
 * Returns null if WASM can't handle.
 */
export function wasmHypotScalar(a: ArrayStorage, scalar: number): ArrayStorage | null {
  if (!a.isCContiguous) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = a.dtype;

  // Float path
  const floatKernel = scalarKernels[dtype];
  if (floatKernel) {
    const Ctor = ctorMap[dtype]!;
    const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
    const outBytes = size * bpe;

    const outRegion = wasmMalloc(outBytes);
    if (!outRegion) return null;

    wasmConfig.wasmCallCount++;

    resetScratchAllocator();

    let aPtr: number;
    if (dtype === 'float16') {
      aPtr = f16InputToScratchF32(a, size);
    } else {
      aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, bpe);
    }

    floatKernel(aPtr, outRegion.ptr, size, scalar);

    if (dtype === 'float16') {
      const f16Region = f32OutputToF16Region(outRegion, size);
      outRegion.release();
      if (!f16Region) return null;
      return ArrayStorage.fromWasmRegion(
        Array.from(a.shape),
        dtype,
        f16Region,
        size,
        Float16Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
      );
    }

    return ArrayStorage.fromWasmRegion(
      Array.from(a.shape),
      dtype,
      outRegion,
      size,
      Ctor as unknown as new (buffer: ArrayBuffer, byteOffset: number, length: number) => TypedArray
    );
  }

  // Int path
  const intKernel = intScalarKernels[dtype];
  if (intKernel) {
    const InputCtor = ctorMap[dtype]!;
    const inputBpe = (InputCtor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
    const outBytes = size * 8;

    const outRegion = wasmMalloc(outBytes);
    if (!outRegion) return null;

    wasmConfig.wasmCallCount++;

    resetScratchAllocator();
    const aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, inputBpe);

    intKernel(aPtr, outRegion.ptr, size, scalar);

    return ArrayStorage.fromWasmRegion(
      Array.from(a.shape),
      'float64',
      outRegion,
      size,
      Float64Array as unknown as new (
        buffer: ArrayBuffer,
        byteOffset: number,
        length: number
      ) => TypedArray
    );
  }

  return null;
}

File: src/common/wasm/zig/hypot.zig
//! WASM element-wise hypotenuse kernels for float types.
//!
//! Binary: out[i] = sqrt(a[i]^2 + b[i]^2)
//! Scalar: out[i] = sqrt(a[i]^2 + scalar^2)
//! Both operate on contiguous 1D buffers of length N.

const simd = @import("simd.zig");
const math = @import("std").math;

/// Element-wise hypotenuse for f64: out[i] = sqrt(a[i]^2 + b[i]^2).
export fn hypot_f64(a: [*]const f64, b: [*]const f64, out: [*]f64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = @sqrt(a[i] * a[i] + b[i] * b[i]);
    }
}

/// Element-wise hypotenuse scalar for f64: out[i] = sqrt(a[i]^2 + scalar^2).
export fn hypot_scalar_f64(a: [*]const f64, out: [*]f64, N: u32, scalar: f64) void {
    const s2 = scalar * scalar;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = @sqrt(a[i] * a[i] + s2);
    }
}

/// Element-wise hypotenuse for f32: out[i] = sqrt(a[i]^2 + b[i]^2).
export fn hypot_f32(a: [*]const f32, b: [*]const f32, out: [*]f32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = @sqrt(a[i] * a[i] + b[i] * b[i]);
    }
}

/// Element-wise hypotenuse scalar for f32: out[i] = sqrt(a[i]^2 + scalar^2).
export fn hypot_scalar_f32(a: [*]const f32, out: [*]f32, N: u32, scalar: f32) void {
    const s2 = scalar * scalar;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        out[i] = @sqrt(a[i] * a[i] + s2);
    }
}

/// Element-wise hypotenuse for i64 → f64 output. Scalar (no i64 SIMD in WASM).
export fn hypot_i64(a: [*]const i64, b: [*]const i64, out: [*]f64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        const bf = @as(f64, @floatFromInt(b[i]));
        out[i] = @sqrt(af * af + bf * bf);
    }
}

/// Element-wise hypotenuse scalar for i64 → f64 output.
export fn hypot_scalar_i64(a: [*]const i64, out: [*]f64, N: u32, scalar: i64) void {
    const sf = @as(f64, @floatFromInt(scalar));
    const s2 = sf * sf;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        out[i] = @sqrt(af * af + s2);
    }
}

/// Element-wise hypotenuse for i32 → f64 output.
/// Uses SIMD: load i32x4, convert to f32x4, compute, widen to f64 for accumulation.
export fn hypot_i32(a: [*]const i32, b: [*]const i32, out: [*]f64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        const bf = @as(f64, @floatFromInt(b[i]));
        out[i] = @sqrt(af * af + bf * bf);
    }
}

/// Element-wise hypotenuse scalar for i32 → f64 output.
export fn hypot_scalar_i32(a: [*]const i32, out: [*]f64, N: u32, scalar: i32) void {
    const sf = @as(f64, @floatFromInt(scalar));
    const s2 = sf * sf;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        out[i] = @sqrt(af * af + s2);
    }
}

/// Element-wise hypotenuse for i16 → f32 output.
export fn hypot_i16(a: [*]const i16, b: [*]const i16, out: [*]f32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        const bf = @as(f64, @floatFromInt(b[i]));
        out[i] = @floatCast(@sqrt(af * af + bf * bf));
    }
}

/// Element-wise hypotenuse scalar for i16 → f32 output.
export fn hypot_scalar_i16(a: [*]const i16, out: [*]f32, N: u32, scalar: i16) void {
    const sf = @as(f64, @floatFromInt(scalar));
    const s2 = sf * sf;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        out[i] = @floatCast(@sqrt(af * af + s2));
    }
}

/// Element-wise hypotenuse for i8 → f32 output.
export fn hypot_i8(a: [*]const i8, b: [*]const i8, out: [*]f32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        const bf = @as(f64, @floatFromInt(b[i]));
        out[i] = @floatCast(@sqrt(af * af + bf * bf));
    }
}

/// Element-wise hypotenuse scalar for i8 → f32 output.
export fn hypot_scalar_i8(a: [*]const i8, out: [*]f32, N: u32, scalar: i8) void {
    const sf = @as(f64, @floatFromInt(scalar));
    const s2 = sf * sf;
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const af = @as(f64, @floatFromInt(a[i]));
        out[i] = @floatCast(@sqrt(af * af + s2));
    }
}

// --- Tests ---

test "hypot_f64 basic (3,4)=5 and (5,12)=13" {
    const testing = @import("std").testing;
    const a = [_]f64{ 3.0, 5.0, 0.0 };
    const b = [_]f64{ 4.0, 12.0, 0.0 };
    var out: [3]f64 = undefined;
    hypot_f64(&a, &b, &out, 3);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 13.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 0.0, 1e-10);
}

test "hypot_f32 basic (3,4)=5 and (5,12)=13" {
    const testing = @import("std").testing;
    const a = [_]f32{ 3.0, 5.0, 0.0 };
    const b = [_]f32{ 4.0, 12.0, 0.0 };
    var out: [3]f32 = undefined;
    hypot_f32(&a, &b, &out, 3);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-5);
    try testing.expectApproxEqAbs(out[1], 13.0, 1e-5);
    try testing.expectApproxEqAbs(out[2], 0.0, 1e-5);
}

test "hypot_scalar_f64 basic" {
    const testing = @import("std").testing;
    const a = [_]f64{ 3.0, 0.0 };
    var out: [2]f64 = undefined;
    hypot_scalar_f64(&a, &out, 2, 4.0);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 4.0, 1e-10);
}

test "hypot_scalar_f32 basic" {
    const testing = @import("std").testing;
    const a = [_]f32{ 3.0, 0.0 };
    var out: [2]f32 = undefined;
    hypot_scalar_f32(&a, &out, 2, 4.0);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-5);
    try testing.expectApproxEqAbs(out[1], 4.0, 1e-5);
}

test "hypot_f64 N=1 boundary" {
    const testing = @import("std").testing;
    const a = [_]f64{5.0};
    const b = [_]f64{12.0};
    var out: [1]f64 = undefined;
    hypot_f64(&a, &b, &out, 1);
    try testing.expectApproxEqAbs(out[0], 13.0, 1e-10);
}

test "hypot_i64 basic" {
    const testing = @import("std").testing;
    const a = [_]i64{ 3, 5, 0 };
    const b = [_]i64{ 4, 12, 0 };
    var out: [3]f64 = undefined;
    hypot_i64(&a, &b, &out, 3);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 13.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 0.0, 1e-10);
}

test "hypot_scalar_i64 basic" {
    const testing = @import("std").testing;
    const a = [_]i64{ 3, 0 };
    var out: [2]f64 = undefined;
    hypot_scalar_i64(&a, &out, 2, 4);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 4.0, 1e-10);
}

test "hypot_i32 basic" {
    const testing = @import("std").testing;
    const a = [_]i32{3};
    const b = [_]i32{4};
    var out: [1]f64 = undefined;
    hypot_i32(&a, &b, &out, 1);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-10);
}

test "hypot_scalar_i32 basic" {
    const testing = @import("std").testing;
    const a = [_]i32{3};
    var out: [1]f64 = undefined;
    hypot_scalar_i32(&a, &out, 1, 4);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-10);
}

test "hypot_i16 basic" {
    const testing = @import("std").testing;
    const a = [_]i16{3};
    const b = [_]i16{4};
    var out: [1]f32 = undefined;
    hypot_i16(&a, &b, &out, 1);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-5);
}

test "hypot_scalar_i16 basic" {
    const testing = @import("std").testing;
    const a = [_]i16{3};
    var out: [1]f32 = undefined;
    hypot_scalar_i16(&a, &out, 1, 4);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-5);
}

test "hypot_i8 basic" {
    const testing = @import("std").testing;
    const a = [_]i8{3};
    const b = [_]i8{4};
    var out: [1]f32 = undefined;
    hypot_i8(&a, &b, &out, 1);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-5);
}

test "hypot_scalar_i8 basic" {
    const testing = @import("std").testing;
    const a = [_]i8{3};
    var out: [1]f32 = undefined;
    hypot_scalar_i8(&a, &out, 1, 4);
    try testing.expectApproxEqAbs(out[0], 5.0, 1e-5);
}
